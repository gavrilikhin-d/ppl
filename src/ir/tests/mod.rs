use crate::test_compilation_result;

test_compilation_result!(empty_constructor);
