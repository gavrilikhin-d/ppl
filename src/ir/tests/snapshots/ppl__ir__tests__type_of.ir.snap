---
source: src/ir/tests/mod.rs
expression: ir
---
; ModuleID = 'type_of'
source_filename = "type_of.ppl"

%"Type<String>" = type { %String, %Integer }
%String = type { ptr }
%Integer = type { ptr }
%"Type<Integer>" = type { %String, %Integer }

@"Type<String>" = private global %"Type<String>" zeroinitializer
@0 = private unnamed_addr constant [7 x i8] c"String\00", align 1
@"Type<Integer>" = private global %"Type<Integer>" zeroinitializer
@1 = private unnamed_addr constant [8 x i8] c"Integer\00", align 1
@ty = global %"Type<Integer>" zeroinitializer

define private void @initialize() !dbg !3 {
  %1 = alloca %"Type<String>", align 8, !dbg !7
  %"Type<String>.name" = getelementptr inbounds %"Type<String>", ptr %1, i32 0, i32 0, !dbg !7
  %2 = call %String @string_from_c_string_and_length(ptr @0, i64 6), !dbg !8
  store %String %2, ptr %"Type<String>.name", align 8, !dbg !8
  %"Type<String>.size" = getelementptr inbounds %"Type<String>", ptr %1, i32 0, i32 1, !dbg !8
  %3 = call %Integer @integer_from_i64(i64 8), !dbg !8
  store %Integer %3, ptr %"Type<String>.size", align 8, !dbg !8
  %4 = load %"Type<String>", ptr %1, align 8, !dbg !8
  store %"Type<String>" %4, ptr @"Type<String>", align 8, !dbg !8
  br label %return, !dbg !8

return:                                           ; preds = %0
  ret void
}

declare %String @string_from_c_string_and_length(ptr, i64)

declare %Integer @integer_from_i64(i64)

define private void @initialize.1() !dbg !9 {
  %1 = alloca %"Type<Integer>", align 8, !dbg !10
  %"Type<Integer>.name" = getelementptr inbounds %"Type<Integer>", ptr %1, i32 0, i32 0, !dbg !10
  %2 = call %String @string_from_c_string_and_length(ptr @1, i64 7), !dbg !11
  store %String %2, ptr %"Type<Integer>.name", align 8, !dbg !11
  %"Type<Integer>.size" = getelementptr inbounds %"Type<Integer>", ptr %1, i32 0, i32 1, !dbg !11
  %3 = call %Integer @integer_from_i64(i64 8), !dbg !11
  store %Integer %3, ptr %"Type<Integer>.size", align 8, !dbg !11
  %4 = load %"Type<Integer>", ptr %1, align 8, !dbg !11
  store %"Type<Integer>" %4, ptr @"Type<Integer>", align 8, !dbg !11
  br label %return, !dbg !11

return:                                           ; preds = %0
  ret void
}

define private void @initialize.2() !dbg !12 {
  %1 = call %Integer @integer_from_i64(i64 1), !dbg !13
  %2 = call %"Type<Integer>" @"type of <:Integer>"(%Integer %1), !dbg !13
  store %"Type<Integer>" %2, ptr @ty, align 8, !dbg !13
  br label %return, !dbg !13

return:                                           ; preds = %0
  ret void
}

define %"Type<Integer>" @"type of <:Integer>"(%Integer %0) !dbg !14 {
  %return_value = alloca %"Type<Integer>", align 8
  %2 = load %"Type<Integer>", ptr @"Type<Integer>", align 8, !dbg !15
  store %"Type<Integer>" %2, ptr %return_value, align 8, !dbg !15
  br label %return, !dbg !15

return:                                           ; preds = %1
  %3 = load %"Type<Integer>", ptr %return_value, align 8
  ret %"Type<Integer>" %3
}

define void @type_of.execute() !dbg !16 {
init_globals:
  call void @initialize(), !dbg !17
  call void @initialize.1(), !dbg !17
  call void @initialize.2(), !dbg !18
  br label %0, !dbg !18

0:                                                ; preds = %init_globals
  %1 = load %"Type<Integer>", ptr @ty, align 8, !dbg !19
  call void @"println <:Type<Integer>>"(%"Type<Integer>" %1), !dbg !19
  br label %return, !dbg !18

return:                                           ; preds = %0
  ret void
}

define void @"println <:Type<Integer>>"(%"Type<Integer>" %0) !dbg !20 {
  %x = alloca %"Type<Integer>", align 8
  store %"Type<Integer>" %0, ptr %x, align 8
  %2 = load %"Type<Integer>", ptr %x, align 8, !dbg !21
  %3 = call %String @"String from <:Type<Integer>>"(%"Type<Integer>" %2), !dbg !21
  call void @"println <:String>"(%String %3), !dbg !21
  br label %return, !dbg !21

return:                                           ; preds = %1
  ret void
}

declare void @"println <:String>"(%String)

define %String @"String from <:Type<Integer>>"(%"Type<Integer>" %0) !dbg !22 {
  %return_value = alloca %String, align 8
  %ty = alloca %"Type<Integer>", align 8
  store %"Type<Integer>" %0, ptr %ty, align 8
  %name = getelementptr inbounds %"Type<Integer>", ptr %ty, i32 0, i32 0, !dbg !23
  %2 = load %String, ptr %name, align 8, !dbg !23
  store %String %2, ptr %return_value, align 8, !dbg !23
  br label %return, !dbg !23

return:                                           ; preds = %1
  %3 = load %String, ptr %return_value, align 8
  ret %String %3
}

!llvm.module.flags = !{!0}
!llvm.dbg.cu = !{!1}

!0 = !{i32 2, !"Debug Info Version", i32 3}
!1 = distinct !DICompileUnit(language: DW_LANG_C, file: !2, producer: "ppl", isOptimized: false, runtimeVersion: 0, emissionKind: FullDebug, splitDebugInlining: false, sysroot: "/")
!2 = !DIFile(filename: "type_of.ppl", directory: ".")
!3 = distinct !DISubprogram(name: "initialize", linkageName: "initialize", scope: !2, file: !2, line: 1, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!4 = !DISubroutineType(types: !5)
!5 = !{!6}
!6 = !DIBasicType(name: "i32", size: 32, encoding: DW_ATE_signed)
!7 = !DILocation(line: 1, column: 10, scope: !3)
!8 = !DILocation(line: 0, scope: !3)
!9 = distinct !DISubprogram(name: "initialize.1", linkageName: "initialize.1", scope: !2, file: !2, line: 1, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!10 = !DILocation(line: 1, column: 10, scope: !9)
!11 = !DILocation(line: 0, scope: !9)
!12 = distinct !DISubprogram(name: "initialize.2", linkageName: "initialize.2", scope: !2, file: !2, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!13 = !DILocation(line: 0, column: 19, scope: !12)
!14 = distinct !DISubprogram(name: "type of <:Integer>", linkageName: "type of <:Integer>", scope: !12, file: !2, line: 1, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!15 = !DILocation(line: 1, column: 10, scope: !14)
!16 = distinct !DISubprogram(name: "type_of.execute", linkageName: "type_of.execute", scope: !2, file: !2, line: 1, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!17 = !DILocation(line: 1, column: 10, scope: !16)
!18 = !DILocation(line: 0, column: 9, scope: !16)
!19 = !DILocation(line: 1, column: 8, scope: !16)
!20 = distinct !DISubprogram(name: "println <:Type<Integer>>", linkageName: "println <:Type<Integer>>", scope: !16, file: !2, line: 1, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!21 = !DILocation(line: 1, column: 10, scope: !20)
!22 = distinct !DISubprogram(name: "String from <:Type<Integer>>", linkageName: "String from <:Type<Integer>>", scope: !20, file: !2, line: 1, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!23 = !DILocation(line: 1, column: 10, scope: !22)
